README.md
# Select Autocomplete

_Version 0.2 ([changelog](CHANGELOG.md))_

A select or input with autocomplete of predefined values.

![](screenshot.gif)

[See Demo](https://csspoo.com/select-autocomplete/example.html)

## In short

- Simple to use
- Small filesize (1.5kB minified + gzipped)
- Keyboard support
- No dependencies
- HTML web component
- Vanilla JS

## Setup

### CSS

Place it within `<head></head>`.

<!-- prettier-ignore -->
```html
<link rel="stylesheet" href="assets/css/dist/select-autocomplete.min.css">
```

### Javaascript

In most cases place it just before `</body>`.

```html
<script src="assets/js/dist/select-autocomplete.min.js"></script>
```

## Usage

<!-- prettier-ignore -->
```html
<select-autocomplete>
  <label><input placeholder="Select a country..."></label>

  <datalist>
    <option value="Finland">
    <option value="Norway">
    <option value="Sweden">
  </datalist>
</select-autocomplete>
```

You can add more HTML to this component, if you need to. However, make sure to not remove the current elements. All of the elements above are are required, except for the `label`.

## Options

There are no options except for what you see in the usage example above.

## Events

### Submit

The `submit` event is fired when the user first select an option and then press enter to confirm.

The example below will add the `submit` event to the first `select-autocomplete` element.

```html
<script>
  document
    .querySelector("select-autocomplete")
    .addEventListener("submit", (e) => {
      console.log(e.detail.value);
    });
</script>
```

### Toggle

The `toggle` event is fired when the user toggles the item list by clicking the input field or the label.

## Compared to input with datalist

There is already a native way to provide autocomplete with combining an input field with a datalist. However there are some benefits of using this script instead.

1. It's much easier to style, because it does not use the shadow dom.
1. It's easier to manipulate the HTML, because it does not use the shadow dom.
1. It behaves a bit different, better in our opinion.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist

## Donate

Donate to [https://www.paypal.me/DevoneraAB](DevoneraAB) if you want.

## License

MIT


CHANGELOG.md
# Changelog

### 0.2

- Added fix for multiple components when clicking outside.

### 0.1

- Initial release


index.php
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Autocomplete input</title>
  <link rel="stylesheet" href="assets/css/select-autocomplete.css?t=<?= time(); ?>">

  <style>
    body {
      font-family: arial;
    }
  </style>
</head>

<body>
  <form action="/action_page.php">
    <select-autocomplete>
      <label><input name="whatever" placeholder="Sök län... "></label>

      <datalist>
        <option value="Norde">
        <option value="abc">
        <option value="aabb">
      </datalist>
    </select-autocomplete>


    <select-autocomplete>
      <label><input placeholder="Sök nåt annat... "></label>

      <datalist>
        <option value="77">
        <option value="55 åäö">
        <option value="78">
      </datalist>
    </select-autocomplete>
  </form>



  Nånting annat

  <script src="assets/js/select-autocomplete.js?t=<?= time(); ?>"></script>
  <script>
    window.addEventListener('DOMContentLoaded', () => {
      document.querySelector('select-autocomplete').addEventListener('submit', (e) => {
        console.log(e.detail);
      });
    });
  </script>
</body>

</html>

gulpfile.js
let gulp = require("gulp");
let sass = require("gulp-sass");
let autoprefixer = require("gulp-autoprefixer");
let concat = require("gulp-concat");
let uglify = require("gulp-uglify-es").default;
let csso = require("gulp-csso");
let rename = require("gulp-rename");

var css = {
  src: "assets/css/src/**/*.scss",
  dest: "assets/css/dist",
  filename: "select-autocomplete.scss",
};

var js = {
  src: "assets/js/src/**/*.js",
  dest: "assets/js/dist",
  filename: "select-autocomplete.js",
};

function style() {
  return gulp
    .src(css.src)
    .pipe(concat(css.filename))
    .pipe(sass())
    .on("error", sass.logError)
    .pipe(autoprefixer())
    .pipe(gulp.dest(css.dest))
    .pipe(csso())
    .pipe(rename({ extname: ".min.css" }))
    .pipe(gulp.dest(css.dest));
}

function script() {
  return gulp
    .src(js.src)
    .pipe(concat(js.filename))
    .pipe(gulp.dest(js.dest))
    .pipe(uglify())
    .pipe(rename({ extname: ".min.js" }))
    .pipe(gulp.dest(js.dest));
}

function watch() {
  gulp.watch(css.src, style);
  gulp.watch(js.src, script);
}

exports.css = style;
exports.js = script;
exports.default = watch;


assets/js/autocomplete-input.js
/*
filtered custom event
input enter custom event
Sätt strong på match

// Custom kod - Clicka pil som en select box
// Outside click event hide all

// Pil icon
// Pil toggle
// Vid klick i select, visa alla likt keydown

// Dölj även wrapper vid dölj och visa den om items finns

// Rensa
// active vid key arrow
// active vid mus

// Fallbacks
// classes
// active class
// placeholder

// Testa multipla element
// Stöd för flera klasser
*/

class AutocompleteInput extends HTMLElement {
  constructor() {
    super();

    console.log("testar");

    this.handleKeydown = this.handleKeydown.bind(this);
    this.handleEnter = this.handleEnter.bind(this);
    this.handleClickItem = this.handleClickItem.bind(this);
    this.handleArrowdown = this.handleArrowdown.bind(this);
    this.handleArrowup = this.handleArrowup.bind(this);
    /*
    this.handleClickItem = this.handleClickItem.bind(this);
    this.handleClickInput = this.handleClickInput.bind(this);*/
  }

  connectedCallback() {
    this.innerHTML = `
      <label>
        <input placeholder="${this.getAttribute("placeholder")}">
      </label>
      <div></div>
    `;

    this.setData();
    this.populate();
    this.cacheInput();
    this.storeItems();
    this.storeList();
    this.storeClassActive();
    this.triggerEvents();
  }

  cacheInput() {
    this.input = this.querySelector("input");
  }

  storeItems() {
    this.items = this.querySelectorAll("div > div");
  }

  storeList() {
    const listSelector = `#${this.getAttribute("list")}`;
    this.listEl = document.querySelector(listSelector);
  }

  storeClassActive() {
    this.classActive = this.listEl.dataset.classActive;
  }

  setData() {
    const list = this.getAttribute("list");
    const options = document.querySelectorAll(`#${list} option`);
    this.data = {};

    options.forEach((option) => {
      this.data[option.value] = option.innerHTML;
    });
  }

  populate() {
    const listSelector = `#${this.getAttribute("list")}`;
    const classes = document.querySelector(listSelector).dataset.classWrap;

    let list = this.querySelector("div");

    list.setAttribute("class", classes);

    for (const value in this.data) {
      const title = this.data[value];
      const item = document.createElement("div");

      item.textContent = title;
      item.setAttribute("hidden", "");
      item.dataset.value = value;
      list.appendChild(item);
    }
  }

  filter() {
    this.items.forEach((el) => {
      const value = el.innerHTML.toLowerCase();
      const input = this.input.value.toLowerCase();

      if (input.length > 0 && value.indexOf(input) > -1) {
        el.removeAttribute("hidden");
      } else {
        el.setAttribute("hidden", "");
      }
    });

    this.activateFirst();
    this.done = false;
  }

  customEventSubmit() {
    this.dispatchEvent(
      new CustomEvent("submit", {
        detail: {
          value: this.input.value,
          title: "",
        },
      })
    );
  }

  hideAll() {
    this.items.forEach((el) => {
      el.setAttribute("hidden", "");
    });
  }

  activate(el) {
    this.deactivateAll();
    el.classList.add(this.classActive);
    this.active = el;
  }

  deactivateAll() {
    this.items.forEach((el) => {
      el.classList.remove(this.classActive);
    });
  }

  activateFirst() {
    this.deactivateAll();

    const root = this.items[0].parentNode;
    const item = root.querySelector("div:not([hidden])");
    if (!item) return;

    item.classList.add(this.classActive);
    this.active = item;

    console.log(this.active);
  }

  activateLast() {
    this.deactivateAll();

    const root = this.items[0].parentNode;
    const item = root.querySelector("div:not([hidden]):last-of-type");
    if (!item) return;

    item.classList.add(this.classActive);
    this.active = item;
  }

  triggerEvents() {
    this.onKeydown();
    this.onEnter();

    this.onClickItem();
    this.onArrowdown();
    this.onArrowup();
    /*
    
    this.onClickInput();*/
  }

  // Event keydown
  onKeydown() {
    this.input.addEventListener("input", this.handleKeydown);
  }

  handleKeydown() {
    this.filter();
    console.log("asdaas");
  }

  // Event enter
  onEnter() {
    this.input.addEventListener("keydown", this.handleEnter);
  }

  handleEnter(e) {
    if (e.keyCode === 13) {
      if (!this.done) {
        this.input.value = this.active.innerHTML;
        this.handleClickEnter();
        this.done = true;
      } else {
        this.customEventSubmit();
      }
    }
  }

  // Event click item
  onClickItem() {
    this.items.forEach((el) => {
      el.addEventListener("click", this.handleClickItem);
    });
  }

  handleClickItem(e) {
    this.input.value = e.currentTarget.innerHTML;
    this.handleClickEnter();
    //this.triggerItemClick();
  }

  handleClickEnter() {
    this.filter();
    this.hideAll();
    this.input.focus();
  }

  // Event arrow down
  onArrowdown() {
    this.input.addEventListener("keydown", this.handleArrowdown);
  }

  handleArrowdown(e) {
    if (e.keyCode === 40) {
      e.preventDefault();

      let result = this.active.nextElementSibling;

      while (result && result.hidden == true) {
        result = result.nextElementSibling;
      }

      if (result) {
        this.activate(result);
      } else {
        this.activateFirst();
      }
    }
  }

  // Event arrow up
  onArrowup() {
    this.input.addEventListener("keydown", this.handleArrowup);
  }

  handleArrowup(e) {
    if (e.keyCode === 38) {
      e.preventDefault();

      let result = this.active.previousElementSibling;

      while (result && result.hidden == true) {
        result = result.previousElementSibling;
      }

      if (result) {
        this.activate(result);
      } else {
        this.activateLast();
      }
    }
  }
}

customElements.define("input-autocomplete", AutocompleteInput);

/*
    this.selectors = Object.assign(this.defaultSelectors(), this.o.selectors);
    this.styles = Object.assign(this.defaultStyles(), this.o.styles);

  onClickInput() {}

  handleKeydown() {
    this.filter();
  }

  handleClickInput(e) {}

  triggerEventInputClick() {}

  triggerItemClick() {
    // VID KLICK EJ KEYDWN
    this.customEventItemClick = new CustomEvent("click:item");
    this.dispatchEvent(this.customEventItemClick);
  }
}
*/


assets/js/select-autocomplete.js
class SelectAutocomplete extends HTMLElement {
  constructor() {
    super();

    this.handleKeydown = this.handleKeydown.bind(this);
    this.handleEnter = this.handleEnter.bind(this);
    this.handleEsc = this.handleEsc.bind(this);
    this.handleClickItem = this.handleClickItem.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.handleClickInput = this.handleClickInput.bind(this);
    this.handleClickLabel = this.handleClickLabel.bind(this);
    this.handleHoverItem = this.handleHoverItem.bind(this);
    this.handleArrowdown = this.handleArrowdown.bind(this);
    this.handleArrowup = this.handleArrowup.bind(this);
  }

  connectedCallback() {
    this.populate();
    this.storage();
    this.querySelector("datalist").remove();
    this.dataset.open = "false";
    this.store.input.setAttribute("autocomplete", "off");
    this.triggerEvents();
  }

  populate() {
    let html = "";

    this.querySelectorAll("option").forEach((el) => {
      html += `<div>${el.value}</div>\n`;
    });

    html = `<div data-list hidden>${html}</div>`;
    this.querySelector("datalist").insertAdjacentHTML("beforebegin", html);
  }

  storage() {
    this.store = {
      values: this.values,
      input: this.querySelector("input"),
      list: this.querySelector("[data-list]"),
      items: this.querySelectorAll("[data-list] > div"),
    };
  }

  filter() {
    this.store.hasVisible = false;

    this.store.items.forEach((el) => {
      const value = el.innerText;
      const input = this.store.input.value;
      const match = value.toLowerCase().indexOf(input.toLowerCase()) > -1;

      if (input !== "" && match) {
        el.removeAttribute("hidden");
        this.store.hasVisible = true;
        el.innerHTML = this.boldString(value, input);
      } else {
        el.setAttribute("hidden", "");
      }
    });

    this.triggerListVisibility();
    this.activateFirst();
    this.triggerMenuDirection();
    this.done = false;
  }

  boldString(str, find) {
    var reg = new RegExp("(" + find + ")", "gi");
    return str.replace(reg, "<strong>$1</strong>");
  }

  triggerListVisibility() {
    if (this.store.hasVisible) {
      this.store.list.removeAttribute("hidden");
    } else {
      this.store.list.setAttribute("hidden", "");
    }
  }

  triggerMenuDirection() {
    if (this.store.hasVisible) {
      this.dataset.open = "true";
    } else {
      this.dataset.open = "false";
    }
  }

  // Custom event submit
  customEventSubmit() {
    this.dispatchEvent(
      new CustomEvent("submit", {
        detail: {
          value: this.store.input.value,
        },
      })
    );
  }

  hideAll() {
    this.store.list.setAttribute("hidden", "");
    this.store.items.forEach((el) => {
      el.setAttribute("hidden", "");
    });
    this.store.hasVisible = false;
    this.triggerMenuDirection();
  }

  showAll() {
    this.deactivateAll();
    this.store.list.removeAttribute("hidden");

    this.store.items.forEach((el) => {
      this.stripTags(el);
      el.removeAttribute("hidden");
    });
    this.store.hasVisible = true;
    this.triggerMenuDirection();
  }

  stripTags(el) {
    el.innerHTML = el.innerText;
  }

  activate(el) {
    this.deactivateAll();
    el.dataset.active = "";
    this.active = el;
  }

  deactivateAll() {
    this.store.items.forEach((el) => {
      delete el.dataset.active;
    });
  }

  activateFirst() {
    const item = this.store.list.querySelector("div:not([hidden])");
    if (item) {
      this.activate(item);
    } else {
      this.active = null;
    }
  }

  activateBySelector(selector) {
    const item = this.store.list.querySelector(selector);
    if (!item) return;
    this.activate(item);
  }

  triggerEvents() {
    this.onKeydown();
    this.onEnter();
    this.onEsc();

    this.onClickInput();
    this.onClickItem();
    this.onClickOutside();
    this.onClickLabel();
    this.onHoverItem();

    this.onArrowdown();
    this.onArrowup();
  }

  // Event keydown
  onKeydown() {
    this.store.input.addEventListener("input", this.handleKeydown);
  }

  handleKeydown() {
    this.filter();
  }

  // Event enter
  onEnter() {
    this.store.input.addEventListener("keydown", this.handleEnter);
  }

  handleEnter(e) {
    if (e.keyCode === 13) {
      if (this.store.input.value == "" && !this.active) {
        this.showAll();
      } else if (!this.done) {
        if (this.active) {
          this.store.input.value = this.active.innerText;
        }
        this.handleClickEnter();
        this.done = true;
      } else {
        this.customEventSubmit();
      }
    }
  }

  // Event enter
  onEsc() {
    this.store.input.addEventListener("keydown", this.handleEsc);
  }

  handleEsc(e) {
    if (e.keyCode === 27) {
      this.hideAll();
    }
  }

  // Event click input
  onClickInput() {
    this.store.input.addEventListener("click", this.handleClickInput);
  }

  handleClickInput(e) {
    e.stopPropagation();

    if (this.store.input.value == "") {
      this.showAll();
    } else {
      this.filter();
    }
  }

  // Event click label
  onClickLabel() {
    this.querySelector("label").addEventListener(
      "click",
      this.handleClickLabel
    );
  }

  handleClickLabel(e) {
    e.preventDefault();

    if (this.dataset.open === "false") {
      if (this.store.input.value == "") {
        this.showAll();
      } else {
        this.filter();
      }
    } else {
      this.hideAll();
    }

    this.store.input.focus();
  }

  // Event click outside
  onClickOutside() {
    window.addEventListener("click", this.handleClickOutside);
  }

  handleClickOutside(e) {
    if (e.target.closest("select-autocomplete")) return;
    this.hideAll();
    this.triggerMenuDirection();
  }

  // Event click item
  onClickItem() {
    this.store.items.forEach((el) => {
      el.addEventListener("click", this.handleClickItem);
    });
  }

  handleClickItem(e) {
    this.store.input.value = e.currentTarget.innerText;
    this.handleClickEnter();
  }

  handleClickEnter() {
    this.filter();
    this.hideAll();
    this.triggerMenuDirection();
    this.store.input.focus();
  }

  // Event hover item
  onHoverItem() {
    this.store.items.forEach((el) => {
      el.addEventListener("mouseover", this.handleHoverItem);
    });
  }

  handleHoverItem(e) {
    this.activate(e.currentTarget);
  }

  // Event arrow
  onArrowup() {
    this.store.input.addEventListener("keydown", this.handleArrowup);
  }

  onArrowdown() {
    this.store.input.addEventListener("keydown", this.handleArrowdown);
  }

  handleArrowup(e) {
    if (e.keyCode === 38 && this.active) {
      const selector = "div:not([hidden]):last-of-type";
      this.handleArrow(e, this.active.previousElementSibling, selector);
    }
  }

  handleArrowdown(e) {
    if (e.keyCode === 40) {
      if (this.active) {
        const selector = "div:not([hidden]):first-of-type";
        this.handleArrow(e, this.active.nextElementSibling, selector);
      } else {
        this.activateFirst();
      }
    }
  }

  handleArrow(e, result, selector) {
    e.preventDefault();

    while (result && result.hidden == true) {
      result =
        e.keyCode == 38
          ? result.previousElementSibling
          : result.nextElementSibling;
    }

    if (result) {
      this.activate(result);
    } else {
      this.activateBySelector(selector);
    }
  }
}

customElements.define("select-autocomplete", SelectAutocomplete);
